### ISSUE TITLE
no space left on device

### ISSUE THREAD

**Issue by antonkulaga** [Azure/azure-storage-fuse#130]
When I copy large files (like unzipped ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/idmapping_selected.tab.gz ) to mounted azore storage I often get:
```
cp: error writing '/storage/indexes/uniprot/idmapping_selected.tab': No space left on device
```
I wonder how can it be possible at all?

**Comment by seguler** [comment]
It is possible. Blobfuse uses a cache directory to store the file when opened. It seems you do not have enough space in your cache directory (--tmp-path=...)

**Comment by kriston** [comment]
I have plenty of space in my cache directory and I get this error when copying files to my mounted Azure blob container.

I worked around this by using fusermount to unmount and then re-mounting the volume, after which the size increased.

It looks like the blobfuse driver needs to pro-actively monitor the size of the mounted container.  At no time did any of the volumes on my local system exceed capacity.  The blobfuse driver told my application that its disk was full, while, at the same time, the blobfuse driver wasn't aware that the container dynamically increases in size as required.

I think this is a valid bug that should be tracked.

**Comment by seguler** [comment]
What do you see when you run ```df -k``` when you get this error ? Can you copy paste the output here for blobfuse ? 

Are you able to consistently repro this issue ? Is there anyway you can provide your repro steps ?

**Comment by kriston** [comment]
It always says this.  Right now this storage account is empty and blobfuse reports 11 GB used and 11 GB free on a 22 GB sized volume.

Filesystem     1K-blocks     Used Available Use% Mounted on
blobfuse        22009344 10843228  11166116  50% /mnt/fileshare

**Comment by seguler** [comment]
We are reporting the underlying disk partition size here. So if you have configured your tmp location to /mnt/blobfusetmp; df -k will report blobfuse file system statistics based on that partition that serves /mnt/blobfusetmp.

I do admit this is confusing. Why does blobfuse not report Blob storage statistics instead ? This is because it is not very straigh-forward to calculate the space used in Blob storage, vs free space etc.

@AlexanderYukhanov @asorrin-msft Any ideas what we should do to resolve this confusion ?

**Comment by kriston** [comment]
When I got the disk full error the blobfusetmp still had 11 GB free, so something weird is going on.

If it helps, this happens during one long rsync operation.  When I stop the operation and restart it, it continues until it hits the error again.

**Comment by seguler** [comment]
Do you have any large files that are 10-11GB ?

**Comment by seguler** [comment]
As a side note - we have a known issue at this moment with rsync: https://github.com/Azure/azure-storage-fuse/issues/74

Rsync uses file rename during data sync. We do not not garbage collect the cache data that has gone through renames... This is a bug in blobfuse code. So if you work on 100GB of data transfer with rsync, you need 100GB in your cache folder... Only unmounting and remounting blobfuse will address this issue since unmounting cleans up the cache folder automatically.

**Comment by kriston** [comment]
We don't have any 11+ GB files but we do have folders larger that is larger than 11 GB.

Thanks for the information and the help.

**Comment by seguler** [comment]
Rsync issue has been resolved in 0.4 release. Closing this.
